### PR TITLE
HBox Component

### PR DESCRIPTION
**Pull request by Zombiefruit**
Added hbox component to wix-ui.

### CODE REVIEW

File: packages/wix-ui-core/src/components/StylableHBox/HBoxStyle.st.css [hunk 1]
```diff
@@ -0,0 +1,11 @@
+:import {
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/StylableHBox/HBoxStyle.st.css]: I think that it is ok to say that this component is style-free so there is no default look. I would remove this file

File: packages/wix-ui-core/src/components/StylableHBox/HBox.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  spacedChildren.push(children[children.length - 1]);
+
+  return spacedChildren;
+};
+
+/**
+ * HBox
+ */
+export const HBox: React.SFC<HBoxProps> = props => {
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/StylableHBox/HBox.tsx]: Let's talk about this. I think if we move the logic of spacing to css, we will not have this wrapping layer

File: packages/wix-ui-core/stories/HBox/HBox-story.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
+import * as React from 'react';
```
Comment (lbelinsk) [inline packages/wix-ui-core/stories/HBox/HBox-story.tsx]: Didn't pull to see the story, but I am not sure why can't we simply use the autodocs. Sure there is a good reason, so we can talk about it tomorrow

File: packages/wix-ui-core/src/components/StylableHBox/HBox.tsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+export interface HBoxProps {
+  children?: React.ReactNode;
+  verticalAlignment?: 'top' | 'center' | 'bottom';
+  spacing?: number;
+}
+
+const addSpacing = (children, spacing) => {
+  if (!children) {
+    return null;
```
[the remark this answers is not in the document]
Comment (lbelinsk) [reply packages/wix-ui-core/src/components/StylableHBox/HBox.tsx]: `if (!children || !spacing) return children;`

File: packages/wix-ui-core/src/components/StylableHBox/HBox.tsx [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  spacedChildren.push(children[children.length - 1]);
+
+  return spacedChildren;
+};
+
+/**
+ * HBox
+ */
+export const HBox: React.SFC<HBoxProps> = props => {
+  const {verticalAlignment, spacing} = props;
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/StylableHBox/HBox.tsx]: props.children

File: packages/wix-ui-core/stories/HBox/HBox-story.tsx (hunk 3 above)
Comment (lbelinsk) [reply packages/wix-ui-core/stories/HBox/HBox-story.tsx]: ask Arijus if needed

File: packages/wix-ui-core/src/components/StylableHBox/HBox.tsx [hunk 6]
```diff
@@ -7,19 +7,25 @@ export interface HBoxProps {
...
     return children;
   }
 
+  const marginDirection = direction === 'right' ? 'marginRight' : direction === 'bottom' ? 'marginBottom' : 'marginRight';
+
   const spacedChildren = [];
   for (let i = 0; i < children.length - 1; i++) {
     spacedChildren.push(
-      <div style={{marginRight: spacing}}>
+      <div style={{marginDirection: spacing}}>
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/StylableHBox/HBox.tsx]: you mean [marginDirection]

Comment (lbelinsk) [reply packages/wix-ui-core/src/components/StylableHBox/HBox.tsx]: I think just define this function a little bit more typed (to get only "buootm" | "right") and then:
```
<div style={{[`margin${direction}`]: spacing}}>
```
To prevent the trinary condition

Comment (Zombiefruit) [reply packages/wix-ui-core/src/components/StylableHBox/HBox.tsx]: ...Yes, my bad

File: packages/wix-ui-core/src/components/StylableHBox/HBox.e2e.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+import {getStoryUrl, waitForVisibilityOf} from 'wix-ui-test-utils/protractor';
+import {hboxTestkitFactory} from '../../testkit/protractor';
+
+describe('HBox', () => {
+  const storyUrl = getStoryUrl('Components', 'HBox');
+
+  beforeEach(() => browser.get(storyUrl));
+  eyes.it('should display correct content', () => {
+    const dataHook = 'storybook-hbox';
+    const driver = hboxTestkitFactory({dataHook});
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/StylableHBox/HBox.e2e.ts]: hBoxTestkitFactory if you ask me, but not critical

File: packages/wix-ui-core/src/components/StylableHBox/HBox.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import * as React from 'react';
+import {mount} from 'enzyme';
+import {HBox} from './';
+
+describe('HBox', () => {
+  let wrapper;
+
+  afterEach(() => wrapper.detach());
+
+  it('should render the passed children', () => {
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/StylableHBox/HBox.spec.tsx]: what about testing the spacing? we can use getComputedStyle
Maybe in eye for eyes is enough?

File: packages/wix-ui-core/src/components/StylableHBox/HBox.st.css [hunk 9]
```diff
@@ -0,0 +1,23 @@
+.root {
+    -st-states: verticalAlignment(enum(bottom, center, top)), rtl;
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/StylableHBox/HBox.st.css]: instead of rtl, let's use `dir(enum(ltr, rtl))`

Comment (lbelinsk) [reply packages/wix-ui-core/src/components/StylableHBox/HBox.st.css]: the reason for it is that we are using it in toggleswitch and slider that way

File: packages/wix-ui-core/src/components/StylableHBox/HBox.st.css [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+.root:verticalAlignment(center) {
+    align-items: center;
+}
+
+.root:verticalAlignment(top) {
+    align-items: flex-start;
+}
+
+.root:rtl {
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/StylableHBox/HBox.st.css]: But if on the body the user will do `<body dir="rtl">` it will be taken care of automatically.
so no need for that actually.

We will assume that the user places a `dir=rtl`

File: packages/wix-ui-core/src/components/StylableHBox/HBox.tsx [hunk 11]
```diff
@@ -0,0 +1,28 @@
+import * as React from 'react';
+import {number, bool, any, oneOf, arrayOf} from 'prop-types';
+import style from './HBox.st.css';
+import {addSpacing} from './utils';
+
+export interface HBoxProps {
+  children?: React.ReactNode;
+  verticalAlignment?: AlignmentOptions;
+  spacing?: number;
+  rtl?: boolean;
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/StylableHBox/HBox.tsx]: `dir?: 'rtl' | 'ltr'`

File: packages/wix-ui-core/src/components/StylableHBox/utils.tsx [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+export function addSpacing (children, spacing: number, rtl?: boolean) {
+  if (!children || !spacing) {
+    return children;
+  }
+
+  const spacedChildren = [];
+
+  if (rtl) {
+    for (let i = children.length - 1; i > 0; i--) {
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/StylableHBox/utils.tsx]: Instead of looping twice:

```js
const marginProperty === rtl ? 'marginLeft' : 'marginRight';
for (let i = children.length - 1; i > 0; i--) {
   spacedChildren.push(
        <div style={{[`$marginProperty`]: spacing}}>
          {children[i]}
        </div>
}
```

File: packages/wix-ui-core/src/components/HBox/HBox.e2e.ts [hunk 13]
```diff
@@ -0,0 +1,16 @@
+import * as eyes from 'eyes.it';
+import {browser} from 'protractor';
+import {getStoryUrl, waitForVisibilityOf} from 'wix-ui-test-utils/protractor';
+import {hboxTestkitFactory} from '../../testkit/protractor';
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/HBox/HBox.e2e.ts]: hBoxTestkitFactory

File: packages/wix-ui-core/src/components/HBox/HBox.protractor.driver.ts [hunk 14]
```diff
@@ -0,0 +1,4 @@
+export const hboxDriverFactory = component => ({
+    /** returns the component element */
+    element: () => component
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/HBox/HBox.protractor.driver.ts]: I would add getChildren as well here

File: packages/wix-ui-core/src/components/HBox/HBox.tsx [hunk 15]
```diff
@@ -1,15 +1,28 @@
...
+  children?: React.ReactNode;
+  verticalAlignment?: AlignmentOptions;
+  spacing?: number;
+  dir?: 'rtl' | 'ltr';
 }
 
-const HBox: React.SFC<HBoxProps> = ({children, classes}) => (
-  <div className={classes.hRoot}>{children}</div>
-);
+export type AlignmentOptions = 'top' | 'center' | 'bottom';
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/HBox/HBox.tsx]: You define the Alignment type here, so i would rename it from AlignmentOptions to Alignment. Up to you

File: packages/wix-ui-core/src/components/HBox/HBox.tsx [hunk 16]
```diff
@@ -1,15 +1,28 @@
...
-export default createHOC(HBox);
+/**
+ * HBox
+ */
+export const HBox: React.SFC<HBoxProps> = props => {
+  const {children, verticalAlignment, spacing, dir} = props;
+  return <div {...style('root', {verticalAlignment}, props)}>{addSpacing(children, spacing, dir)}</div>;
+};
+
+HBox.propTypes = {
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/HBox/HBox.tsx]: what about defaultProps?

File: packages/wix-ui-core/src/components/HBox/utils.tsx [hunk 17]
```diff
@@ -0,0 +1,22 @@
+import * as React from 'react';
+
+export function addSpacing (children, spacing: number, dir?: 'rtl' | 'ltr') {
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/HBox/utils.tsx]: Maybe worth to define `type Direction = 'ltr' | 'rtl'` 
It is being used several times

File: packages/wix-ui-core/src/components/HBox/utils.tsx [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+  if (!children || !spacing) {
+    return children;
+  }
+
+  const spacedChildren = [];
+  const marginProperty = dir === 'rtl' ? 'marginLeft' : 'marginRight';
+
+  for (let i = 0; i < children.length - 1; i++) {
+    spacedChildren.push(
+      <div style={{[marginProperty]: spacing}}>
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/HBox/utils.tsx]: You are adding extra wrapping here for each child...
I think it will be better (less bad) to use react.cloneElement and add the margin to it, without an extra div

File: packages/wix-ui-core/src/testkit/protractor.ts [hunk 19]
```diff
@@ -35,6 +35,9 @@ export const stylablebadgeTestkitFactory = protractorTestkitFactoryCreator(styla
 import {autocompleteDriverFactory} from '../components/Autocomplete/Autocomplete.protractor.driver';
 export const autocompleteTestkitFactory = protractorTestkitFactoryCreator(autocompleteDriverFactory);
 
+import {hboxDriverFactory} from '../components/HBox/HBox.protractor.driver';
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/testkit/protractor.ts]: hBox

Comment (lbelinsk) [reply packages/wix-ui-core/src/testkit/protractor.ts]: Also, expose the testkit from the other `index` and `enzyme` files

File: packages/wix-ui-core/src/components/HBox/HBox.protractor.driver.ts [hunk 20]
```diff
@@ -1,4 +1,3 @@
-export const hboxDriverFactory = component => ({
-    /** returns the component element */
-    element: () => component
-  });
+export const hBoxDriverFactory = component => ({
+  element: () => component
```
Comment (lbelinsk) [inline packages/wix-ui-core/src/components/HBox/HBox.protractor.driver.ts]: why did you remove the comments?
